sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -0,0 +1,285 @@
+      extends PTransform<PCollection<String>, PCollection<KV<String, String>>> {
+
+    @Nullable abstract RedisConnectionConfiguration connectionConfiguration();
+
+    abstract ReadAll.Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      @Nullable abstract ReadAll.Builder setConnectionConfiguration(
+          RedisConnectionConfiguration connection);
+      abstract ReadAll build();
+    }
+
+    public ReadAll withConnectionConfiguration(RedisConnectionConfiguration connection) {
+      checkArgument(connection != null, "connection can not be null");
+      return builder().setConnectionConfiguration(connection).build();
+    }
+
+    @Override
+    public PCollection<KV<String, String>> expand(PCollection<String> input) {
+      checkArgument(connectionConfiguration() != null,
+          "withConnectionConfiguration() is required");
+
+      return input
+          .apply(ParDo.of(
+              new ReadFn(connectionConfiguration())
+          ))
+          .setCoder(KvCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()))
+          .apply(new Reparallelize());
+    }
+
+  }
+
+  /**
+   * A {@link DoFn} requesting Redis server to get key/value pairs.
+   */
+  private static class ReadFn extends DoFn<String, KV<String, String>> {
+
+    private final RedisConnectionConfiguration connectionConfiguration;
+
+    private transient Jedis jedis;
+
+    public ReadFn(RedisConnectionConfiguration connectionConfiguration) {
+      this.connectionConfiguration = connectionConfiguration;
+    }
+
+    @Setup
+    public void setup() {
+      jedis = connectionConfiguration.connect();
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext processContext) throws Exception {
+      ScanParams scanParams = new ScanParams();
+      scanParams.match(processContext.element());
+
+      String cursor = ScanParams.SCAN_POINTER_START;
+      boolean finished = false;
+      while (!finished) {
+        ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
+        List<String> keys = scanResult.getResult();
+
+        Pipeline pipeline = jedis.pipelined();
+        if (keys != null) {
+          for (String key : keys) {
+            pipeline.get(key);
+          }
+          List<Object> values = pipeline.syncAndReturnAll();
+          for (int i = 0; i < values.size(); i++) {
+            processContext.output(KV.of(keys.get(i), (String) values.get(i)));
+          }
+        }
+
+        cursor = scanResult.getStringCursor();
+        if (cursor.equals("0")) {
+          finished = true;
+        }
+      }
+    }
+
+    @Teardown
+    public void teardown() {
+      jedis.close();
+    }
+
+  }
+
+  private static class Reparallelize
+      extends PTransform<PCollection<KV<String, String>>, PCollection<KV<String, String>>> {
+
+    @Override public PCollection<KV<String, String>> expand(PCollection<KV<String, String>> input) {
+      // breaking fusion
```

Comment:
I suggest to explicitly say here that this is copied from JdbcIO with the same caveats, because this code is making some subtle assumptions and it's good to make it traceable to the source of the assumptions.

Also, consider rebasing on https://github.com/apache/beam/pull/3890 and using Reshuffle.viaRandomKey() when that goes in.